### ISSUE TITLE
미세조정 학습 가이드 문의

### ISSUE THREAD

**Issue by jehoon-ed** [AIRC-KETI/ke-t5#2]
안녕하세요,
좋은 자료를 공유해 주셔서 정말 감사드립니다.

pytorch huggingface 모델을 이용하여 간단한 conversation task를 실험하고 있습니다.
미세조정 실험에 있어, optimizer 그리고 scheduler에 대한 가이드를 받고 싶어 문의 드립니다.

추가로, t5-base 모델과 t5-large 모델을 실험하고 있으며,
(https://discuss.huggingface.co/t/t5-finetuning-tips/684/3) 를 참고하여 optimizer를 아래와 같이 설정하였습니다.

- Adafactor(model.parameters(), scale_parameter=True, relative_step=True, warmup_init=True, lr=None)
- Adafactor(model.parameters(), scale_parameter=False, relative_step=False, warmup_init=False, lr=1e-3)

Base/Large 모델 모두 설정한 validation set에서 학습이 지속됨에 따라 loss가 급격하게 상승하는 현상을 보이고 있습니다.
특히, large 모델에서는 validation set에서 굉장히 큰 loss를 보이며, 실제 학습된 모델을 통해 inference 시 start token(pad token) 다음 토큰으로 EOS 토큰을 생성하여 정상적인 문장 생성에 차질을 겪고 있습니다.

이러한 상황에서 정상적인 문장 생성을 위해 시도해봐야 하는 것이 무엇이 있을지 가이드 부탁 드립니다.

좋은 자료로 연구에 도움을 주신 점 다시 한번 정말 감사드립니다 :)

**Comment by kimsan0622** [comment]
말씀하신 상황의 정확한 원인은 잘 모르겠습니다.
저희가 huggingface T5로 downstream tasks 학습시킬때는 딱히 문제가 발생하지 않아서요.

일단 Adafactor에 대해서 말씀드리면, huggingface의 adafactor impl을 까보면 알아서 inverse square root scheduling을 하도록 되어있습니다. 즉, relative_step = True면 T5에서 사전학습을 진행할때와 같이 10,000 step까지는 1e-2의 learning rate로 학습되고 그  이후에는 1/sqrt(step)으로 learning rate가 schedule됩니다. learning rate schedule을 직접하고 싶으시면 직접 optimizer의 learning rate를 매 step또는 epoch마다 바꿔주시면 됩니다. huggingface adafactor optimizer에 learning rate를 설정하면 constant learning rate만 됩니다. 아마 T5논문에서 training에서는 invsqrt learning rate scheduling하고, downstreams에서는 constant learning rate를 이용해서 저렇게 구현한거 같습니다.

일단 https://huggingface.co/transformers/main_classes/optimizer_schedules.html#transformers.Adafactor 여길 참고하셔서,

- Adafactor(model.parameters(), scale_parameter=False, relative_step=False, warmup_init=False, lr=1e-3)
- Adafactor(model.parameters(), scale_parameter=True, relative_step=True, warmup_init=True, lr=None)
- # replace AdamW with Adafactor
- optimizer = Adafactor(

    model.parameters(),
    lr=1e-3,
    eps=(1e-30, 1e-3),
    clip_threshold=1.0,
    decay_rate=-0.8,
    beta1=None,
    weight_decay=0.0,
    relative_step=False,
    scale_parameter=False,
    warmup_init=False
)

위 configuration들로 한번 진행해 보시기 바랍니다.
저희가 해본 결과로는 AdamW나 Adafactor나 다 잘 학습되는것을 확인했습니다.

그리고 inference하실때 generate로 num_beams 랑 repetition_penalty, min_length, max_length 등등을 조절해보시기 바랍니다. 저 같은 경우 dialogue modeling에서는 보통 num_beams는 4, temperature 1.0, top_k 10, top_p 0.9, length_penalty 0.65 정도 씁니다. repetition_penalty는 쓸때도 있고 안쓸때도 있는데, 쓸때는 2.0정도 줍니다.

**Comment by jehoon-ed** [comment]
친절하고 상세한 답변 감사합니다. 
말씀주신 training, inference 설정으로 테스트를 진행해 보겠습니다.

혹시, t5-large 모델을 학습할 때, 메모리 문제로 인해 굉장히 작은 batch size를 설정하여 학습을 진행하였습니다. (batch_size=3)
이렇게 작은 배치사이즈로 인한 문제일 가능성도 있을지 문의 드립니다.

**Comment by kimsan0622** [comment]
가능성이 없진 않을거 같습니다. DDP와 Dynamic Batching(https://parl.ai/docs/tutorial_worlds.html#dynamic-batching)으로 배치사이즈를 늘려보시는 것도 방법일 것 같습니다.

하지만 데이터의 영향이 더 클거 같은데요.

만약 학습하시는 데이터가 open-domain dialogue modeling이시라면 대답들이 너무 발산할 수도 있을 것 같습니다. 대답할 수 있는게 대답이 너무 다양한거죠... 그래서 GPT1-3처럼 auto regressive objective로 학습시킨 후 살짝 fine-tuning하는 것도 도움이 될 수 있을거 같습니다. 이것은 그냥 raw corpus이용해서 huggingface t5 transformer에 input_ids랑 labels에 같은 passage를 넣어서 0번째 output인 loss를 학습시키시면 됩니다. 그러나 장담할 수 없습니다.
 
따라서 condition들을 좀 늘릴 필요가 있을거 같습니다.
1. dialogue가 multi-turn일 경우 dialogue history 전체를 context로 넣는다. single-turn이면 dialogue history가 없어서 불가능하겠죠 ㅠㅠ
2. 할 수 있는 모든 condition들을 찾는다. intent라던지, 아니면 https://github.com/facebookresearch/EmpatheticDialogues 처럼 dialogue의 감정을 넣는다던지, image chat처럼 image를 넣는다던지, wizard of wikipedia처럼 information retrieval을 통해 knowledge를 넣는다던지...

제가 multi-turn dialogue에서 dialogue history를 몽땅 때려넣었을 때는 그럭저럭 결과가 나왔었습니다. 뭐, 역시나 자주 등장하는 대답 패턴들이 꽤 보이긴 했지만요... GPT3처럼 AR 마구 돌려서 쓰면 https://arxiv.org/pdf/2104.07567.pdf 에 의하면 plausible한 거짓말들을 마구 생성해낸다고 합니다. 루머 생성기여도 괜찮으시면 그 방법을 사용하셔도 괜찮을거 같습니다.

아직 딥러닝 모델을 이용한 dialogue에서의 response generation은 전세계적으로 만족할만한 성능이 나오고 있진 않는거 같으니, 이런 저런 실험을 하시다보면 재밌는 결과들이 나올 거 같습니다. 다만 application 만들어서 서비스하실거면 response를 generation하지 말고 selection하시는걸 추천드립니다.

**Comment by jehoon-ed** [comment]
답변 감사합니다.

동일한 input에 대해 서로 다른 label이 태깅되어 있는 데이터들이 존재하고, 이로인해 추가의 condition 없이 학습이 지속되다 보니, 계속 발산하는 결과가 나온 것 같습니다.

GPT2 모델도 함께 테스트를 진행해보고 있는데, 말씀하신 대로 dialogue task에서는 GPT2의 결과가 정성적으로 미세하지만 약간은 좋은 것으로 보여집니다. (제가 학습을 제대로 하지 못하여 얻은 결과일 가능성도 있습니다)

https://arxiv.org/pdf/2104.07567.pdf 논문은 아직 접하지 못하였는데 흥미로운 논문일 것 같습니다. 참고하겠습니다.

여러 reference와 함께 훌륭한 가이드를 해주신 덕분에 어느정도 진행해 볼 수 있는 방향이 생긴 것 같습니다.
이슈 해결과 함께 재미있는 결과가 나온다면 여러 분들이 참고하실 수 있게 여기에 공유드릴 수 있도록 하겠습니다.

감사합니다. 👍
